llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -79,187 +88,241 @@ const Constant *getAliasee(const GlobalAlias *GA) {
-// contributions.
-template <typename Function>
-static bool updateDigestUseDependenciesAndContributions(
-    const GlobalObject *GO, MD5 &GOHash, GOStateMap &Visited, GOInfoMap &GOIMap,
-    Function AccumulateGODigest) {
-  if (GO->isDeclaration())
-    return false;
+auto ModuleHashGenerator::updateDigestUseDependenciesAndContributions(
+    const GlobalObject *GO, bool HasRepoDefinitionMD)
+    -> std::tuple<size_t, DigestType> {
+  assert(!GO->isDeclaration() && "Can only be used for global definitions");
+
+  const auto GODepth = Visited.size();
+  LLVM_DEBUG(dbgs() << "Computing hash for \"" << GO->getName() << "\" (#"
+                    << GODepth << ")\n");
+
+  // If the hash has been memoized, we can return the result immediately.
+  bool Hit;
+  DigestType GODigest;
+  std::tie(GODigest, Hit) = getGOHash(GO);
+  if (Hit) {
+    ++VisitedCachedGOTimes;
+    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
+                      << "\"\n");
+    return std::make_tuple(GODepth, GODigest);
+  }
 
+  MD5 GOHash = MD5();
   bool Inserted;
   typename GOStateMap::const_iterator StateIt;
-  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, Visited.size());
+  // Have we previously visited this GO on this path? If so, add to the hash a
+  // back-reference to that GO and return its position to the caller. If not,
+  // record that we have visited this GO and its depth. This will be used to
+  // form a back-reference if we loop back here in future.
+  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, GODepth);
   if (!Inserted) {
-    // If GO is visited, use the letter 'R' as the marker and use its state as
-    // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    assert(GODepth > StateIt->second);
+    // If GO is visited, mark as a Backref and use its state as the value.
+    GOHash.update(static_cast<char>(Tags::Backref));
+    GOHash.update(GODepth - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    DigestType Digest;
+    GOHash.final(Digest);
+    Changed = true;
+    return std::make_tuple(StateIt->second, Digest);
   }
 
-  GOHash.update('T');
-  auto State = AccumulateGODigest(GO, GOHash, GOIMap);
-  bool Changed = State.Changed;
+  GOHash.update(static_cast<char>(Tags::GO));
+  auto State = accumulateGODigest(GO, GOHash, HasRepoDefinitionMD);
   // The hashes of the GO's 'contributions' and 'dependencies' are separately
   // added to the GO's hash since their respective arrows point are in different
   // direction. In addition, 'contributions' cannot form loops since 1) only
   // global variables have 'contributions' and 2) all 'contributions' are
   // functions. Therefore, the 'contributions' can be simplified not to record
   // 'visited' table, which avoids having two 'visited' maps. Firstly, add the
```

Comment:
Is this mention of visited tables and maps confusing now?

Reply:
Changed to use `map`.